### PR TITLE
Adding a possibility to refresh Google auth token

### PR DESCRIPTION
**Pull request by carbon-wpa**

### CODE REVIEW

**Comment by jaibatrik** [comment]
Hi, thanks for the PR! I have couple of concern / question / suggestion. Will add them to individual files.

File: projects/lib/src/entities/login-provider.ts
Comment (jaibatrik) [inline projects/lib/src/entities/login-provider.ts:5]: Could you please make it an options object (similar to `signInOptions` in the `signIn` method)?

File: projects/lib/src/socialauth.service.ts
Comment (jaibatrik) [inline projects/lib/src/socialauth.service.ts:104]: I think we can have the `getLoginStatus` method exposed through this, rather than just `refreshAuthToken`.

File: src/app/demo/demo.component.html
Comment (jaibatrik) [inline src/app/demo/demo.component.html:31]: Don't really think this needs to be shown in the demo page. If we do, we can show this only when a user is logged in with Google.

**Review by jaibatrik — CHANGES_REQUESTED** [review]

**Comment by jaibatrik** [comment]
Thanks again for the contribution!

File: projects/lib/src/providers/google-login-provider.ts
Comment (BenjaminWang1031) [inline projects/lib/src/providers/google-login-provider.ts:55]: Hey there, the reloadAuthResponse returns a promise other than a general response. I've done a little bit tweaking locally and it works fine.

      ```
  let user: GoogleUser = new GoogleUser();
        const profile = this.auth2.currentUser.get().getBasicProfile();
        user.id = profile.getId();
        user.name = profile.getName();
        user.email = profile.getEmail();
        user.photoUrl = profile.getImageUrl();
        user.firstName = profile.getGivenName();
        user.lastName = profile.getFamilyName();
        user.response = profile;

        if(refreshToken)
        {
            this.auth2.currentUser.get()
                .reloadAuthResponse(true)
                .then(response=>{
                    user.authToken = response.access_token;
                    user.idToken = response.id_token;
                    resolve(user);
                })
        }
        else
        {
            const authResponse = this.auth2.currentUser.get().getAuthResponse(true);
            user.authToken = authResponse.access_token;
            user.idToken = authResponse.id_token;
            resolve(user);
        }
```

Comment (BenjaminWang1031) [reply projects/lib/src/providers/google-login-provider.ts:55]: @jaibatrik @carbon-wpa Btw, thank you guys for making this happen! Since token is valid only for an hour, this feature is really helpful to not interrupt operations or upset people. We can put a setting in our apps for the valid login duration and build some logic around it to refresh the token when needed.

File: projects/lib/src/entities/login-provider.ts
Comment (carbon-wpa) [reply projects/lib/src/entities/login-provider.ts]: I applied similar approach (please check it in the new commit).

File: projects/lib/src/socialauth.service.ts
Comment (carbon-wpa) [reply projects/lib/src/socialauth.service.ts]: Good point. Now  `refreshAuthToken` is returning `void` as a promise and broadcast next with new `_authState`.

File: src/app/demo/demo.component.html
Comment (carbon-wpa) [reply src/app/demo/demo.component.html]: I have applied `ngIf` to show it only when user is logged by Google provider.

**Comment by carbon-wpa** [comment]
> Thanks again for the contribution!

I'm glad I can help community. By the way I also need this method in my project :D

File: projects/lib/src/providers/google-login-provider.ts
Comment (carbon-wpa) [reply projects/lib/src/providers/google-login-provider.ts:55]: I have applied small corrections to the code and waiting for the feedback 👍

**Comment by carbon-wpa** [comment]
@jaibatrik once PR will be accepted let me squash all commits into one. Right now during a checking process I think it is more readable to see the difference in separate commits.

**Comment by carbon-wpa** [comment]
@jaibatrik, @BenjaminWang1031 Can we make a progress with PR :) ?

**Comment by zrelja** [comment]
@jaibatrik any update on this?

I am eager to see @carbon-wpa in production! :D

**Comment by carbon-wpa** [comment]
> @jaibatrik any update on this?
> 
> I am eager to see @carbon-wpa in production! :D

me too :)

**Comment by carbon-wpa** [comment]
@jaibatrik can we move forward this PR, please? I need this functionality in my project and I don't want to use forked version...

**Comment by jaibatrik** [comment]
Sorry, have been busy lately. I'll take a look and publish. Today or tomorrow, maximum by the weekend.

File: projects/lib/src/providers/google-login-provider.ts
Comment (jaibatrik) [reply projects/lib/src/providers/google-login-provider.ts:55]: This looks good to me. Although, I won't place the check of whether a provider supports refreshing token at `SocialAuthService` level. I'll probably leave that to individual providers. I'll make that refactor later.

**Comment by thangaraja** [comment]
just wanted to know the current status of this issue? can we provide the details on how to use refresh token and how to get the new token after refresh?

**Comment by carbon-wpa** [comment]
@thangaraja here is a simple example of code how you can handle it:

`
   private refreshTokenHandler; // this a handler for timeout
   private refreshTokenSubscription: Subscription; // it think it is not neccesarry while you put this code in app.component


   // in app.component you can set it for exmple in 'ngOnInit'
   public ngOnInit(): void {
    this.refreshTokenHandler = setInterval(() => {
      if (this.appService.getLoggedUser()) { // I am checking if user logged in, once user logged in I am setting a flag in global service in the app
        this.socialAuthService.refreshAuthToken(GoogleLoginProvider.PROVIDER_ID); // refreshing token
      }
    }, 1000 * 60 * 59); //refresh token every 59 minutes

    // once token will be refresh an 'authState' will be broadcasted
    this.refreshTokenSubscription = this.socialAuthService.authState
      .subscribe((user) => {
         // this the way I handle user athorization in my app - using gloabl service. This part could look diffrent in your case.
        this.appService.setIsUserLoggedIn(user !== null);
        this.appService.setLoggedUser(user);
      });
  }

// additionaly you should provide the 'ngOnDestroy' for utilization refreshTokenHandler & refreshTokenSubscription  

`

**Pull request merged**
